# Strobealign min/max evaluation

This repository documents how to reproduce the min/max evaluation experiments
in the paper

"Designing efficient randstrobes for sequence similarity
analyses" by Karami et al., 2023.

(See "Implementing c_max in strobealign" in the appendix of that
paper.)


## Summary

To create a randstrobe, a syncmer is combined with a second one downstream.
This is done in a pseudorandom way by applying a function to the two syncmers
and choosing the combination that *minimizes* the output of that function.
An alternative is to *maximize* the function, but that gives similar results.

Here, we test what happens when running both the min and max version
and combining the results.

The provided snakemake workflow (`Snakefile`) runs the evaluation "from scratch",
that is, no further input files need to be provided.

It does the following:
- Download genome reference files
- Simulate reads (using mason)
- Map reads with BWA-MEM
- Download and compile the two "min" and "max" strobealign versions
- Map reads with the two strobealign versions
- Combine min and max results
- Compute accuracy
- Create the result table in LaTeX format (see `table.tex`)


## Datasets

The datasets used here are based on the datasets created for the following paper:

Sahlin, K. Strobealign:
flexible seed size enables ultra-fast and accurate read alignment.
*Genome Biology 23, 260 (2022)*.
https://doi.org/10.1186/s13059-022-02831-7

See the Supplementary Information in that paper, section
"Note A: Simulations" for a description.
The datasets are the simulated drosophila, maize, CHM13 and rye reads
(these use the "SIM3" parameter settings for the read simulator).

See also the original snakemake workflow file that documents the evaluation done
for the Genome Biology paper:
https://github.com/ksahlin/alignment_evaluation/blob/master/evaluation/Snakefile


### The number of reads has been reduced

The datasets in the Genome Biology paper contain 10 million simulated reads.
Here, we use only 1 million reads.

Note that there is a small difference between the results found by the workflow here
and the numbers reported in the Karami et al. paper
because two different ways were used to arrive at the smaller dataset:

- For the Karami et al. paper,
  the original 10 million simulated reads in each file were truncated to the
  first 1 million.
- The workflow in this repository simulates 1 million reads.

These datasets are not the same because the output generated by mason depends
on all input parameters,
and the number of reads it shall generate is one of these parameters.

To get the results to match exactly:

* Set `N_READS` in the `Snakefile` to 10'000'000
* Add a rule to truncate the generated data files to 1 million reads before
  mapping.

We have not made this modification as it adds hundreds of CPU hours for little gain.


## Running the min/max evaluation


    conda env create --file environment.yml
    conda activate strobealign-eval
    snakemake --cores=all

The final results are in `table.tex`.


## Issues

* `mason_variator` may fail with this error when running on maize or CHM13:

      /home/osboxes/seqan/apps/mason2/mason_variator.cpp:1161
      Assertion failed : snpRecord.getPos() != svRecord.getPos()
      should be true but was 0 (Should be generated non-overlapping (snp pos = 229421492, sv pos = 229421492).)

  This can be solved by using a self-compiled version of mason without assertions.



## Commits

The strobealign commits used are:

- min: https://github.com/ksahlin/strobealign/commit/3223dc5946d9f38814e25a25149548dd146cc8d0 (original)
- max: https://github.com/ksahlin/strobealign/commit/6a837431f29fc3be3c3a74bea507538d9ea5abe7 (tag: min-max)


## Rules for combining reads

### Single-end reads

- If one read is unmapped, take the other
- If one read has lower alignment score, take the other

## Paired-end reads

- If one pair has lower sum of alignment scores, take the other
- If one pair is not marked as proper, take the other
- If one pair has fewer mapped reads, take the other
